### ISSUE TITLE
Suppress follow-up errors in type checking

### ISSUE THREAD

**Issue by flodiebold** [rust-lang/rust-analyzer#6975]
In this situation:
```rust
fn foo() {
    let x = something_broken;
    x.max(1);
}
```
we currently emit a `mismatched-arg-count` diagnostic because we resolve the `max` call to `Iterator::max`. We should avoid emitting diagnostics in such a situation, because `x` was already of an unknown type.

Doing this isn't easy though and requires some refactoring, so this issue is unactionable for now.

**Comment by lnicola** [comment]
Just curious, how are we resolving `max` on an unknown type? Is `chalk` doing that?

**Comment by bjorn3** [comment]
No, that is something rust-analyzer does. Chalk only handles trait solving, not method resolution.

**Comment by lnicola** [comment]
So does this happen for inherent methods too, or only for trait methods? It seems like something is saying "we don't know the type of `x`, but `max` must be `Iterator::max` because it's in scope".

**Comment by flodiebold** [comment]
During method resolution, we ask Chalk 'does this unknown type implement that trait?' and Chalk says 'maybe', so we select the method.

One easy step that would fix most of the issues already would be to check in the argument count diagnostic whether the receiver is unknown.

**Comment by lnicola** [comment]
Would it make sense to take "maybe" as "no"? I'm not sure how often that happens, but in the long run I'd assume that's what the compiler will do.

**Comment by flodiebold** [comment]
No, at that point we need to accept 'maybe'; the compiler does the same. When the method is selected, an obligation is registered to make sure it later becomes a 'yes'.

**Comment by flodiebold** [comment]
> One easy step that would fix most of the issues already would be to check in the argument count diagnostic whether the receiver is unknown.

Did that: #7125

**Comment by grinapo** [comment]
I'm not sure this is the same but I get `mismatched-arg-count` for this (possibly stupid) fragment:
```rust
       let r = rows.next(); // rusqlite::Rows
       match r { 
            Some(r) => match r {
               Ok(r) => {
                    println!( "{:20} {:<10}  {:>20}→{:>20} {:15} {:15}⇒{:15} {:15} {:20} (c{})",  
                        r.get::<usize,String>(5), r.get::<usize,String>(1), …
```
while it compiles fine.

**Comment by flodiebold** [comment]
I assume you get the error on the `get` calls? What type does RA show for `r`?

**Comment by grinapo** [comment]
The bold parts are complained at: 
**r.get::<usize**,String>(5), **r.get::<usize**,String>(1), 

r is a `Row` according to the popup in vscode.

**Comment by flodiebold** [comment]
That's weird. Can you make a new issue, ideally with a small reproduction?

**Comment by Veykril** [comment]
Is this still relevant? The initial snippet no longer triggers a diagnostic for me

**Comment by flodiebold** [comment]
I think the general topic is still relevant.

**Comment by Ten0** [comment]
Since the recent update I'm getting:
```
no field `xxx` on type `&{unknown}`
```
errors. Shouldn't we silence such errors when the type involves `unknown`?

**Comment by lnicola** [comment]
Can we close this after #16583 and the previous changes?

**Comment by Veykril** [comment]
I'd say yes, there are probaly more occasions we should look at but those will pop up when we find them
